### PR TITLE
feat: refactor validation api

### PR DESCRIPTION
**Pull request by SHession**
## What does this change?

This PR changes how the validation logic is surfaced to the consumer. With this PR it is no longer returned from`getElementFromNode` but instead a function is provided called `elementDataValidator` which can be used to validate element values.

## How to test

I have added some additional unit tests to ensure the functionality is working correctly, but should be a no-op for the demo.

## How can we measure success?

Consumers have an easy way of validating element values, whilst validation logic is nicely encapsulated within the element spec.

## Accessibility

- [ ] [Tested with screen reader](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/558398)
- [ ] [Navigable with keyboard](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/39894d)
- [ ] [Colour contrast passed](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/92b913)
- [x] [The change doesn't use only colour to convey meaning](https://www.theguardian.design/2a1e5182b/p/6691bb-accessibility/t/29032f)

### CODE REVIEW

File: src/plugin/__tests__/element.spec.ts
Comment (jonathonherbert) [inline src/plugin/__tests__/element.spec.ts:679]: Super minor: do we need to round trip this data through a node and out again? Could be a smaller test if we just pass object values directly. (We could also make this change in the test below.)

File: src/plugin/helpers/test.ts
Comment (jonathonherbert) [inline src/plugin/helpers/test.ts:84]: One thought on naming – functions are verbs, not nouns – what do you think about `validateElementData`, which describes what this function does, rather than what it is?

File: src/plugin/helpers/element.ts
Comment (jonathonherbert) [inline src/plugin/helpers/element.ts:134]: Torn on typesafety here. I think we do expect our data to be a valid shape for the element, even if it doesn't pass validation – that is to say, our validators expect an object that matches the type provided by the Element, even if the data itself is invalid.

Suspect that if keys are missing, for example, our validators will throw errors. So a lack of typesafety here could lead to runtime problems.

We could use a signature of the kind used in `insertElement` to make our `values` typesafe rather than `unknown`. I'd value your thoughts!

**Review by jonathonherbert — COMMENTED** [review]
Two non-blocking comments and one thought about typesafety, but this is looking good! I'll wait for some feedback re: the latter before approving.

File: src/plugin/__tests__/element.spec.ts
Comment (SHession) [reply src/plugin/__tests__/element.spec.ts:679]: I think I might add a specific test case for this, to ensure that we can always pass the output from `getElementDataFromNode` directly to the validator. Unless you see this as redundant?

Comment (jonathonherbert) [reply src/plugin/__tests__/element.spec.ts]: It just expands the surface area of what we're testing, but I see the intent – no strong opinions 👍

File: src/plugin/helpers/element.ts
Comment (SHession) [reply src/plugin/helpers/element.ts]: Worked through this offline, it's now ready for re-review.

**Review by jonathonherbert — APPROVED** [review]
Ace, tests look good too. There's some repetition in a test name that's worth addressing before we merge, otherwise, 👍

File: src/plugin/__tests__/element.spec.ts
Comment (jonathonherbert) [inline src/plugin/__tests__/element.spec.ts:703]: Guessing this should be renamed.

**Pull request merged**
